File: svc-position/src/test/java/com/beancounter/position/TestMarketValuesWithFx.java
package com.beancounter.position;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.model.Asset;
import com.beancounter.common.model.Currency;
import com.beancounter.common.model.FxRate;
import com.beancounter.common.model.IsoCurrencyPair;
import com.beancounter.common.model.MarketData;
import com.beancounter.common.model.MoneyValues;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.model.Position;
import com.beancounter.common.model.Positions;
import com.beancounter.common.model.PriceData;
import com.beancounter.common.model.Trn;
import com.beancounter.common.model.TrnType;
import com.beancounter.common.utils.AssetUtils;
import com.beancounter.common.utils.MathUtils;
import com.beancounter.common.utils.PortfolioUtils;
import com.beancounter.position.accumulation.AccumulationStrategy;
import com.beancounter.position.accumulation.BuyBehaviour;
import com.beancounter.position.accumulation.SellBehaviour;
import com.beancounter.position.valuation.Gains;
import com.beancounter.position.valuation.MarketValue;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import org.junit.jupiter.api.Test;

class TestMarketValuesWithFx {

  @Test
  void is_MarketValue() {
    Asset asset = AssetUtils.getAsset("Test", "ABC");
    BigDecimal simpleRate = new BigDecimal("0.20");

    Trn buyTrn = new Trn(TrnType.BUY, asset);
    buyTrn.setTradeAmount(new BigDecimal("2000.00"));
    buyTrn.setTradeCurrency(asset.getMarket().getCurrency());
    buyTrn.setTradePortfolioRate(simpleRate);
    buyTrn.setQuantity(new BigDecimal("100"));

    AccumulationStrategy buyBehaviour = new BuyBehaviour();

    Position position = new Position(asset);
    Portfolio portfolio = PortfolioUtils.getPortfolio("MV");
    buyBehaviour.accumulate(buyTrn, portfolio, position);
    Positions positions = new Positions(portfolio);
    positions.add(position);

    MarketData marketData = new MarketData(asset);
    marketData.setClose(new BigDecimal("10.00"));
    marketData.setPreviousClose(new BigDecimal("5.00"));


    // Revalue based on marketData prices
    MoneyValues targetValues = new MoneyValues(new Currency("USD"));
    targetValues.setPriceData(PriceData.Companion.of(marketData));
    targetValues.setAverageCost(new BigDecimal("20.00"));
    targetValues.setPurchases(buyTrn.getTradeAmount());
    targetValues.setCostBasis(buyTrn.getTradeAmount());
    targetValues.setCostValue(buyTrn.getTradeAmount());
    targetValues.setTotalGain(new BigDecimal("-1000.00"));
    targetValues.setUnrealisedGain(new BigDecimal("-1000.00"));
    targetValues.setMarketValue(Objects.requireNonNull(
        MathUtils.multiply(buyTrn.getQuantity(), marketData.getClose())));

    Map<IsoCurrencyPair, FxRate> fxRateMap = getRates(portfolio, asset, simpleRate);
    new MarketValue(new Gains()).value(positions, marketData, fxRateMap);

    assertThat(position.getMoneyValues(Position.In.TRADE))
        .isEqualToIgnoringGivenFields(targetValues, "priceData", "portfolio");

    MoneyValues baseValues = new MoneyValues(new Currency("USD"));
    baseValues.setAverageCost(new BigDecimal("20.00"));
    baseValues.setPriceData(PriceData.Companion.of(marketData));
    baseValues.setPurchases(buyTrn.getTradeAmount());
    baseValues.setCostBasis(buyTrn.getTradeAmount());
    baseValues.setCostValue(buyTrn.getTradeAmount());
    baseValues.setTotalGain(new BigDecimal("-1000.00"));
    baseValues.setUnrealisedGain(new BigDecimal("-1000.00"));
    baseValues.setMarketValue(
        MathUtils.nullSafe(MathUtils.multiply(buyTrn.getQuantity(), marketData.getClose())));

    assertThat(position.getMoneyValues(Position.In.BASE))
        .isEqualToIgnoringGivenFields(baseValues, "priceData", "portfolio");
    MoneyValues pfValues = new MoneyValues(portfolio.getCurrency());
    pfValues.setCostBasis(new BigDecimal("10000.00"));
    pfValues.setPurchases(new BigDecimal("10000.00"));
    pfValues.setPriceData(PriceData.Companion.of(marketData, simpleRate));
    pfValues.setMarketValue(new BigDecimal("200.00"));
    pfValues.setAverageCost(new BigDecimal("100.00"));
    pfValues.setCostValue(new BigDecimal("10000.00"));
    pfValues.setUnrealisedGain(new BigDecimal("-9800.00"));
    pfValues.setTotalGain(new BigDecimal("-9800.00"));

    assertThat(position.getMoneyValues(Position.In.PORTFOLIO))
        .isEqualToIgnoringGivenFields(pfValues, "priceData");

  }

  @Test
  void is_GainsOnSell() {
    Portfolio portfolio = PortfolioUtils.getPortfolio("MV");

    Asset asset = AssetUtils.getAsset("Test", "ABC");
    Map<IsoCurrencyPair, FxRate> fxRateMap = new HashMap<>();

    BigDecimal simpleRate = new BigDecimal("0.20");

    fxRateMap.put(
        IsoCurrencyPair.toPair(
            portfolio.getCurrency(),
            asset.getMarket().getCurrency()),
        new FxRate(
            new Currency("X"), new Currency("X"),
            simpleRate, null));

    Trn buyTrn = new Trn(TrnType.BUY, asset);
    buyTrn.setTradeAmount(new BigDecimal("2000.00"));
    buyTrn.setTradeCurrency(asset.getMarket().getCurrency());
    buyTrn.setTradePortfolioRate(simpleRate);
    buyTrn.setQuantity(new BigDecimal("100"));

    AccumulationStrategy buyBehaviour = new BuyBehaviour();

    Position position = new Position(asset);
    buyBehaviour.accumulate(buyTrn, portfolio, position);
    Positions positions = new Positions(portfolio);
    positions.add(position);

    Trn sellTrn = new Trn(TrnType.SELL, asset);
    sellTrn.setTradeAmount(new BigDecimal("3000.00"));
    sellTrn.setTradeCurrency(asset.getMarket().getCurrency());
    sellTrn.setTradePortfolioRate(simpleRate);
    sellTrn.setQuantity(new BigDecimal("100"));

    AccumulationStrategy sellBehaviour = new SellBehaviour();
    sellBehaviour.accumulate(sellTrn, portfolio, position);
    MarketData marketData = new MarketData(asset);
    marketData.setClose(new BigDecimal("10.00"));
    marketData.setPreviousClose(new BigDecimal("9.00"));

    new MarketValue(new Gains()).value(positions, marketData, fxRateMap);

    MoneyValues usdValues = new MoneyValues(new Currency("USD"));
    usdValues.setMarketValue(new BigDecimal("0"));
    usdValues.setAverageCost(BigDecimal.ZERO);
    usdValues.setPriceData(PriceData.Companion.of(marketData, BigDecimal.ONE));
    usdValues.setPurchases(buyTrn.getTradeAmount());
    usdValues.setSales(sellTrn.getTradeAmount());
    usdValues.setCostValue(BigDecimal.ZERO);
    usdValues.setRealisedGain(new BigDecimal("1000.00"));
    usdValues.setUnrealisedGain(BigDecimal.ZERO);
    usdValues.setTotalGain(new BigDecimal("1000.00"));

    assertThat(position.getMoneyValues(Position.In.TRADE))
        .isEqualToIgnoringGivenFields(usdValues, "priceData", "portfolio");

    assertThat(position.getMoneyValues(Position.In.BASE))
        .isEqualToIgnoringGivenFields(usdValues, "priceData", "portfolio");

    MoneyValues pfValues = new MoneyValues(portfolio.getCurrency());
    pfValues.setMarketValue(new BigDecimal("0"));
    pfValues.setAverageCost(BigDecimal.ZERO);
    pfValues.setPriceData(PriceData.Companion.of(marketData, simpleRate));
    pfValues.setPurchases(
        Objects.requireNonNull(MathUtils.divide(buyTrn.getTradeAmount(), simpleRate)));
    pfValues.setCostValue(BigDecimal.ZERO);
    pfValues.setSales(MathUtils.divide(sellTrn.getTradeAmount(), simpleRate));
    pfValues.setRealisedGain(new BigDecimal("5000.00"));
    pfValues.setUnrealisedGain(BigDecimal.ZERO);
    pfValues.setTotalGain(new BigDecimal("5000.00"));

    assertThat(position.getMoneyValues(Position.In.PORTFOLIO))
        .isEqualToIgnoringGivenFields(pfValues, "priceData", "portfolio");
  }

  @Test
  void is_MarketValueWithNoPriceComputed() {

    Asset asset = AssetUtils.getAsset("Test", "ABC");
    BigDecimal simpleRate = new BigDecimal("0.20");

    Trn buyTrn = new Trn(TrnType.BUY, asset);
    buyTrn.setTradeAmount(new BigDecimal("2000.00"));
    buyTrn.setTradeCurrency(asset.getMarket().getCurrency());
    buyTrn.setTradePortfolioRate(simpleRate);
    buyTrn.setQuantity(new BigDecimal("100"));

    AccumulationStrategy buyBehaviour = new BuyBehaviour();

    Position position = new Position(asset);
    Portfolio portfolio = PortfolioUtils.getPortfolio("MV");
    buyBehaviour.accumulate(buyTrn, portfolio, position);
    Positions positions = new Positions(portfolio);
    positions.add(position);

    MarketData marketData = new MarketData(asset);

    Map<IsoCurrencyPair, FxRate> fxRateMap = getRates(portfolio, asset, simpleRate);

    // Revalue based on No Market data
    Position result = new MarketValue(new Gains()).value(positions, marketData, fxRateMap);
    assertThat(result).isNotNull().hasFieldOrProperty("moneyValues");

  }

  private Map<IsoCurrencyPair, FxRate> getRates(
      Portfolio portfolio,
      Asset asset,
      BigDecimal simpleRate) {
    Map<IsoCurrencyPair, FxRate> fxRateMap = new HashMap<>();

    fxRateMap.put(
        IsoCurrencyPair.toPair(
            portfolio.getCurrency(),
            asset.getMarket().getCurrency()
        ),
        new FxRate(new Currency("test"), new Currency("TEST"),
            simpleRate, null));
    return fxRateMap;
  }
}

File: svc-position/src/test/java/com/beancounter/position/TestMarketValuesWithFx.kt
package com.beancounter.position

import com.beancounter.common.model.*
import com.beancounter.common.model.Currency
import com.beancounter.common.model.IsoCurrencyPair.Companion.toPair
import com.beancounter.common.model.PriceData.Companion.of
import com.beancounter.common.utils.AssetUtils.Companion.getAsset
import com.beancounter.common.utils.MathUtils.Companion.divide
import com.beancounter.common.utils.MathUtils.Companion.multiply
import com.beancounter.common.utils.MathUtils.Companion.nullSafe
import com.beancounter.common.utils.PortfolioUtils.Companion.getPortfolio
import com.beancounter.position.accumulation.AccumulationStrategy
import com.beancounter.position.accumulation.BuyBehaviour
import com.beancounter.position.accumulation.SellBehaviour
import com.beancounter.position.valuation.Gains
import com.beancounter.position.valuation.MarketValue
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.util.*

internal class TestMarketValuesWithFx {
    @Test
    fun is_MarketValue() {
        val asset = getAsset("Test", "ABC")
        val simpleRate = BigDecimal("0.20")
        val buyTrn = Trn(TrnType.BUY, asset, BigDecimal(100))
        buyTrn.tradeAmount = BigDecimal("2000.00")
        buyTrn.tradePortfolioRate = simpleRate
        val buyBehaviour: AccumulationStrategy = BuyBehaviour()
        val position = Position(asset)
        val portfolio = getPortfolio("MV")
        buyBehaviour.accumulate(buyTrn, portfolio, position)
        val positions = Positions(portfolio)
        positions.add(position)
        val marketData = MarketData(asset)
        marketData.close = BigDecimal("10.00")
        marketData.previousClose = BigDecimal("5.00")


        // Revalue based on marketData prices
        val targetValues = MoneyValues(Currency("USD"))
        targetValues.priceData = of(marketData)
        targetValues.averageCost = BigDecimal("20.00")
        targetValues.purchases = buyTrn.tradeAmount
        targetValues.costBasis = buyTrn.tradeAmount
        targetValues.costValue = buyTrn.tradeAmount
        targetValues.totalGain = BigDecimal("-1000.00")
        targetValues.unrealisedGain = BigDecimal("-1000.00")
        targetValues.marketValue = Objects.requireNonNull(
                multiply(buyTrn.quantity, marketData.close))!!
        val fxRateMap = getRates(portfolio, asset, simpleRate)
        MarketValue(Gains()).value(positions, marketData, fxRateMap)
        assertThat(position.getMoneyValues(Position.In.TRADE, position.asset.market.currency))
                .isEqualToIgnoringGivenFields(targetValues, "priceData", "portfolio")
        val baseValues = MoneyValues(Currency("USD"))
        baseValues.averageCost = BigDecimal("20.00")
        baseValues.priceData = of(marketData)
        baseValues.purchases = buyTrn.tradeAmount
        baseValues.costBasis = buyTrn.tradeAmount
        baseValues.costValue = buyTrn.tradeAmount
        baseValues.totalGain = BigDecimal("-1000.00")
        baseValues.unrealisedGain = BigDecimal("-1000.00")
        baseValues.marketValue = nullSafe(multiply(buyTrn.quantity, marketData.close))
        assertThat(position.getMoneyValues(Position.In.BASE, positions.portfolio.base))
                .isEqualToIgnoringGivenFields(baseValues, "priceData", "portfolio")
        val pfValues = MoneyValues(portfolio.currency)
        pfValues.costBasis = BigDecimal("10000.00")
        pfValues.purchases = BigDecimal("10000.00")
        pfValues.priceData = of(marketData, simpleRate)
        pfValues.marketValue = BigDecimal("200.00")
        pfValues.averageCost = BigDecimal("100.00")
        pfValues.costValue = BigDecimal("10000.00")
        pfValues.unrealisedGain = BigDecimal("-9800.00")
        pfValues.totalGain = BigDecimal("-9800.00")
        assertThat(position.getMoneyValues(Position.In.PORTFOLIO, positions.portfolio.currency))
                .isEqualToIgnoringGivenFields(pfValues, "priceData")
    }

    @Test
    fun is_GainsOnSell() {
        val portfolio = getPortfolio("MV")
        val asset = getAsset("Test", "ABC")
        val fxRateMap: MutableMap<IsoCurrencyPair, FxRate> = HashMap()
        val simpleRate = BigDecimal("0.20")
        val pair = toPair(
                portfolio.currency,
                asset.market.currency)
        if (pair != null) {
            fxRateMap[pair] = FxRate(
                    Currency("X"), Currency("X"),
                    simpleRate, null)
        }
        val buyTrn = Trn(TrnType.BUY, asset, BigDecimal(100))
        buyTrn.tradeAmount = BigDecimal("2000.00")
        buyTrn.tradePortfolioRate = simpleRate
        val buyBehaviour: AccumulationStrategy = BuyBehaviour()
        val position = Position(asset)
        buyBehaviour.accumulate(buyTrn, portfolio, position)
        val positions = Positions(portfolio)
        positions.add(position)
        val sellTrn = Trn(TrnType.SELL, asset, BigDecimal(100))
        sellTrn.tradeAmount = BigDecimal("3000.00")
        sellTrn.tradePortfolioRate = simpleRate
        val sellBehaviour: AccumulationStrategy = SellBehaviour()
        sellBehaviour.accumulate(sellTrn, portfolio, position)
        val marketData = MarketData(asset)
        marketData.close = BigDecimal("10.00")
        marketData.previousClose = BigDecimal("9.00")
        MarketValue(Gains()).value(positions, marketData, fxRateMap)
        val usdValues = MoneyValues(Currency("USD"))
        usdValues.marketValue = BigDecimal("0")
        usdValues.averageCost = BigDecimal.ZERO
        usdValues.priceData = of(marketData, BigDecimal.ONE)
        usdValues.purchases = buyTrn.tradeAmount
        usdValues.sales = sellTrn.tradeAmount
        usdValues.costValue = BigDecimal.ZERO
        usdValues.realisedGain = BigDecimal("1000.00")
        usdValues.unrealisedGain = BigDecimal.ZERO
        usdValues.totalGain = BigDecimal("1000.00")
        assertThat(position.getMoneyValues(Position.In.TRADE, position.asset.market.currency))
                .isEqualToIgnoringGivenFields(usdValues, "priceData", "portfolio")
        assertThat(position.getMoneyValues(Position.In.BASE, positions.portfolio.base))
                .isEqualToIgnoringGivenFields(usdValues, "priceData", "portfolio")
        val pfValues = MoneyValues(portfolio.currency)
        pfValues.marketValue = BigDecimal("0")
        pfValues.averageCost = BigDecimal.ZERO
        pfValues.priceData = of(marketData, simpleRate)
        pfValues.purchases = Objects.requireNonNull(divide(buyTrn.tradeAmount, simpleRate))!!
        pfValues.costValue = BigDecimal.ZERO
        pfValues.sales = divide(sellTrn.tradeAmount, simpleRate)!!
        pfValues.realisedGain = BigDecimal("5000.00")
        pfValues.unrealisedGain = BigDecimal.ZERO
        pfValues.totalGain = BigDecimal("5000.00")
        assertThat(position.getMoneyValues(Position.In.PORTFOLIO, positions.portfolio.currency))
                .isEqualToIgnoringGivenFields(pfValues, "priceData", "portfolio")
    }

    @Test
    fun is_MarketValueWithNoPriceComputed() {
        val asset = getAsset("Test", "ABC")
        val simpleRate = BigDecimal("0.20")
        val buyTrn = Trn(TrnType.BUY, asset, BigDecimal(100))
        buyTrn.tradeAmount = BigDecimal("2000.00")
        assertThat(buyTrn.tradeCurrency.code).isEqualTo(asset.market.currency.code)
        buyTrn.tradePortfolioRate = simpleRate
        val buyBehaviour: AccumulationStrategy = BuyBehaviour()
        val position = Position(asset)
        val portfolio = getPortfolio("MV")
        buyBehaviour.accumulate(buyTrn, portfolio, position)
        val positions = Positions(portfolio)
        positions.add(position)
        val marketData = MarketData(asset)
        val fxRateMap = getRates(portfolio, asset, simpleRate)

        // Revalue based on No Market data
        val result = MarketValue(Gains()).value(positions, marketData, fxRateMap)
        assertThat(result).isNotNull.hasFieldOrProperty("moneyValues")
    }

    private fun getRates(
            portfolio: Portfolio,
            asset: Asset,
            simpleRate: BigDecimal): Map<IsoCurrencyPair, FxRate> {
        val fxRateMap: MutableMap<IsoCurrencyPair, FxRate> = HashMap()
        val pair = toPair(
                portfolio.currency,
                asset.market.currency)
        if (pair != null){
            fxRateMap[pair] = FxRate(Currency("test"), Currency("TEST"),
                    simpleRate, null)
        }
        return fxRateMap
    }
}